pkg/virt-api/webhooks/mutating-webhook/mutating-webhook.go

```diff
@@ -0,0 +1,124 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2018 Red Hat, Inc.
+ *
+ */
+
+package mutating_webhook
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+
+	v1beta1 "k8s.io/api/admission/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+
+	"kubevirt.io/kubevirt/pkg/api/v1"
+	kubev1 "kubevirt.io/kubevirt/pkg/api/v1"
+	"kubevirt.io/kubevirt/pkg/log"
+	"kubevirt.io/kubevirt/pkg/virt-api/webhooks"
+)
+
+type patchOperation struct {
+	Op    string      `json:"op"`
+	Path  string      `json:"path"`
+	Value interface{} `json:"value,omitempty"`
+}
+
+type mutateFunc func(*v1beta1.AdmissionReview) *v1beta1.AdmissionResponse
+
+func serve(resp http.ResponseWriter, req *http.Request, mutate mutateFunc) {
+	response := v1beta1.AdmissionReview{}
+	review, err := webhooks.GetAdmissionReview(req)
+
+	if err != nil {
+		resp.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	reviewResponse := mutate(review)
+	if reviewResponse != nil {
+		response.Response = reviewResponse
+		response.Response.UID = review.Request.UID
+	}
+	// reset the Object and OldObject, they are not needed in a response.
+	review.Request.Object = runtime.RawExtension{}
+	review.Request.OldObject = runtime.RawExtension{}
+
+	responseBytes, err := json.Marshal(response)
+	if err != nil {
+		log.Log.Reason(err).Errorf("failed json encode webhook response")
+		resp.WriteHeader(http.StatusBadRequest)
+		return
+	}
+	if _, err := resp.Write(responseBytes); err != nil {
+		log.Log.Reason(err).Errorf("failed to write webhook response")
+		resp.WriteHeader(http.StatusBadRequest)
+		return
+	}
+	resp.WriteHeader(http.StatusOK)
+}
+
+func mutateVMIs(ar *v1beta1.AdmissionReview) *v1beta1.AdmissionResponse {
+	vmiResource := metav1.GroupVersionResource{
+		Group:    v1.VirtualMachineInstanceGroupVersionKind.Group,
+		Version:  v1.VirtualMachineInstanceGroupVersionKind.Version,
+		Resource: "virtualmachineinstances",
+	}
+	if ar.Request.Resource != vmiResource {
+		err := fmt.Errorf("expect resource to be '%s'", vmiResource.Resource)
+		return webhooks.ToAdmissionResponseError(err)
+	}
+
+	raw := ar.Request.Object.Raw
+	vmi := v1.VirtualMachineInstance{}
+
+	err := json.Unmarshal(raw, &vmi)
+	if err != nil {
+		return webhooks.ToAdmissionResponseError(err)
+	}
+
+	// Set VMI defaults
+	kubev1.SetObjectDefaults_VirtualMachineInstance(&vmi)
+
+	var patch []patchOperation
+	var value interface{}
+	value = vmi.Spec
```

Comment:
Do we have any guarantee that the Mutation admission controller runs before the Validation one? If not, it's possible we should be running the resulting spec through the Validation logic at the end of this.

Reply:
Yes. It's guaranteed that the mutations run before the validation.

Reply:
@davidvossel Yes, we have guarantees(good for us :smile: )
https://docs.okd.io/latest/architecture/additional_concepts/dynamic_admission_controllers.html